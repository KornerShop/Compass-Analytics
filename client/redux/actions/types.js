export const TOGGLE_AUTHENTICATED = 'UPDATE_AUTHENTICATED';
export const TOGGLE_FETCHING = 'UPDATE_FETCHING';
export const UPDATE_ERROR_MESSAGE = 'UPDATE_ERROR_MESSAGE';
export const POPULATE_LANG_DATA = 'POPULATE_LANG_DATA';
export const POPULATE_OFFICE_DATA = 'POPULATE_OFFICE_DATA';
export const POPULATE_NAV_DATA = 'POPULATE_NAV_DATA';
export const POPULATE_ZIP_DATA = 'POPULATE_ZIP_DATA';
