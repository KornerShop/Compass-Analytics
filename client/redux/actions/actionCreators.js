import {
  toggleAuthenticated,
  toggleFetching,
  updateErrorMessage,
  populateLangData,
  populateOfficeData,
  populateNavData,
  populateZipData,
} from './actions';

export const requestLogin = (dispatch, getState) => {
  const { fetching, authenticated } = getState();
  fetching === false && dispatch(toggleFetching(true));
  authenticated === true && dispatch(toggleAuthenticated(false));
};

export const receiveLogin = (dispatch, getState) => {
  const { fetching, authenticated } = getState();
  fetching === true && dispatch(toggleFetching(false));
  authenticated === false && dispatch(toggleAuthenticated(true));
};

export const loginError = (dispatch, getState, errorMessage) => {
  const { fetching, authenticated } = getState();
  fetching === true && dispatch(toggleFetching(false));
  authenticated === true && dispatch(toggleAuthenticated(false));
  dispatch(updateErrorMessage(errorMessage));
};

export const requestLogout = (dispatch, getState) => {
  const { fetching, authenticated } = getState();
  fetching === false && dispatch(toggleFetching(true));
  authenticated === false && dispatch(toggleAuthenticated(true));
};

export const receiveLogout = (dispatch, getState) => {
  const { fetching, authenticated } = getState();
  fetching == true && dispatch(toggleFetching(false));
  authenticated === true && dispatch(toggleAuthenticated(false));
};

export const updateLangData = (dispatch, langData) =>
  dispatch(populateLangData(langData));

export const updateOfficeData = (dispatch, officeData) =>
  dispatch(populateOfficeData(officeData));

export const updateNavData = (dispatch, navData) =>
  dispatch(populateNavData(navData));

export const updateZipData = (dispatch, zipData) =>
  dispatch(populateZipData(zipData));

export const loginUser = ({ username, password }) => async (
  dispatch,
  getState,
) => {
  requestLogin(dispatch, getState);
  try {
    const response = await fetch('/users/login', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        username,
        password,
      }),
    });
    if (response.status >= 200 && response.status < 300) {
      const { token } = await response.json();
      localStorage.setItem('token', token);
      receiveLogin(dispatch, getState);
    } else {
      loginError(dispatch, getState, 'Wrong credentials');
    }
  } catch (err) {
    console.error(err);
  }
};

export const verifyToken = () => async (dispatch, getState) => {
  requestLogin(dispatch, getState);
  const token = localStorage.getItem('token');
  if (token) {
    try {
      const response = await fetch('/users/verify', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${token}`,
        },
      });
      if (response.status >= 200 && response.status < 300) {
        const { token: { newToken } } = await response.json();
        localStorage.setItem('token', newToken);
        dispatch(receiveLogin());
        receiveLogin(dispatch, getState);
      } else {
        loginError(dispatch, getState, response.statusText);
      }
    } catch (err) {
      console.error(err);
    }
  }
  loginError(dispatch, getState, 'No token found');
};

export const logoutUser = () => (dispatch, getState) => {
  requestLogout(dispatch, getState);
  localStorage.removeItem('token');
  receiveLogout(dispatch, getState);
};

export const listenForChartData = socket => dispatch => {
  socket.on('populate-lang-data', langData =>
    updateLangData(dispatch, langData),
  );
  socket.on('populate-office-data', officeData =>
    updateOfficeData(dispatch, officeData),
  );
  socket.on('populate-Nav-data', navData =>
    updateNavData(dispatch, navData),
  );
  socket.on('populate-zip-data', zipData =>
    updateOfficeData(dispatch, zipData),
  );
};

export const resetErrorMessage = () => dispatch => dispatch(updateErrorMessage(''));
