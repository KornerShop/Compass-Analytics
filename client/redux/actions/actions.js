/* eslint-disable no-console */

import * as types from './types';

import { config } from '../../utilities/fetch';

export const requestLogin = () => ({
  type: types.LOGIN_REQUEST,
  fetching: true,
  authenticated: false,
});

export const receiveLogin = () => ({
  type: types.LOGIN_SUCCESS,
  fetching: false,
  authenticated: true,
});

export const loginError = errorMessage => ({
  type: types.LOGIN_FAILURE,
  fetching: false,
  authenticated: false,
  errorMessage,
});

export const loginUser = ({ username, password }) => async dispatch => {
  dispatch(requestLogin());
  try {
    const response = await fetch('/auth/login', config('POST', {
      username,
      password
    }));
    if (response.status >= 200 && response.status < 300) {
      const { token } = await response.json();
      localStorage.setItem('token', token);
      dispatch(receiveLogin());
    } else {
      dispatch(loginError(response.statusText));
    }
  } catch(err) {
    console.error(err)
  }
};

export const requestLogout = () => ({
  type: types.LOGOUT_REQUEST,
  fetching: true,
  authenticated: true,
});

export const receiveLogout = () => ({
  type: types.LOGOUT_SUCCESS,
  fetching: false,
  authenticated: false,
});

export const logoutUser = () => dispatch => {
  dispatch(requestLogout());
  localStorage.removeItem('token');
  dispatch(receiveLogout());
};
