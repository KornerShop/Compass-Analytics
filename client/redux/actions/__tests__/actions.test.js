import * as actions from '../actions';

describe('actions', () => {
  test('toggleAuthenticated', () => {
    expect(actions.toggleAuthenticated(true)).toMatchSnapshot();
  });
  test('toggleFetching', () => {
    expect(actions.toggleFetching(true)).toMatchSnapshot();
  });
  test('updateErrorMessage', () => {
    expect(actions.updateErrorMessage('No user with the given username')).toMatchSnapshot();
  });
  test('populateLangData', () => {
    expect(actions.populateLangData([
      {
        _id: "English",
        value: 15
      },
      {
        _id: "Spanish",
        value: 27
      }
    ]
  )).toMatchSnapshot();
  });
  test('populateOfficeData', () => {
    expect(actions.populateOfficeData([{
      _id: {
        office: 'SNAP',
        date: '7/29/2017',
      },
      count: 3,
    }])).toMatchSnapshot();
  });
  test('populateNavData', () => {
    expect(actions.populateNavData([
      {
        _id: {
          office: 'SNAP',
          date: '7/29/2017',
        },
        count: 3,
      }
    ])).toMatchSnapshot();
  });
  test('populateZipData', () => {
    expect(actions.populateZipData([
      {
        _id: 95404,
        count: 2
      }
    ])).toMatchSnapshot();
  });
});
