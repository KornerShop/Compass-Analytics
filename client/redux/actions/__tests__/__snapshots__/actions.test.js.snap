// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions populateLangData 1`] = `
Object {
  "payload": "English",
  "type": "POPULATE_LANG_DATA",
}
`;

exports[`actions populateNavData 1`] = `
Object {
  "payload": Object {
    "_id": Object {
      "date": "7/29/2017",
      "office": "SNAP",
    },
    "count": 3,
  },
  "type": "POPULATE_NAV_DATA",
}
`;

exports[`actions populateOfficeData 1`] = `
Object {
  "payload": Object {
    "_id": Object {
      "date": "7/29/2017",
      "office": "SNAP",
    },
    "count": 3,
  },
  "type": "POPULATE_OFFICE_DATA",
}
`;

exports[`actions populateZipData 1`] = `
Object {
  "payload": Object {
    "_id": 95404,
    "count": 2,
  },
  "type": "POPULATE_ZIP_DATA",
}
`;

exports[`actions toggleAuthenticated 1`] = `
Object {
  "payload": true,
  "type": "UPDATE_AUTHENTICATED",
}
`;

exports[`actions toggleFetching 1`] = `
Object {
  "payload": true,
  "type": "UPDATE_FETCHING",
}
`;

exports[`actions updateErrorMessage 1`] = `
Object {
  "payload": "No user with the given username",
  "type": "UPDATE_ERROR_MESSAGE",
}
`;
